File: LeetCodeCPlusPlus/py57.cpp
// 给出一个无重叠，按照区间起始端点排序的区间列表，在列表中插入一个新的区间，确保仍然有序，不重叠
#include <vector>
using namespace std;
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> result;
        int i = 0;
        while (i<intervals.size() && intervals[i][1] < newInterval[0]) { // 之前的有序部分插入
            result.push_back(intervals[i]);
            i++;
        }
        while (i<intervals.size() && intervals[i][0] <= newInterval[1]) { // 那个需要合并的区间
            newInterval[0] = min(newInterval[0], intervals[i][0]);
            newInterval[1] = max(newInterval[1], intervals[i][1]);
            i++;
        }
        result.push_back(newInterval);
        while(i<intervals.size()) // 剩下的部分
            result.push_back(intervals[i++]);
        return result;
    }
};

File: LeetCodePython/py57.py
# 插入区间，无重叠，按区间起始端点排序的区间插入一个新区间后仍然要求有序不重叠
# 下面的解法是将插入节点与剩下部分混合判断, 要修改，左端点比原来的小，左端点比原来的大，插入后后端点正好又是某个端点的起点
# 也可以分成三部分，前面不需要插入的，中间插入改变的区间的起始结束点，剩下的部分
from typing import List
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        result = list()
        if not intervals:
            return [newInterval]
        index = 0
        n = len(intervals)
        while index<n and intervals[index][1] < newInterval[0]: # 找到可以插入的位置
            result.append(intervals[index])
            index += 1
        if index == n:  # 已经是结尾，直接插入到末尾 
            result.append(newInterval)
            return result
        if index == 0 and intervals[index][0] > newInterval[1] or newInterval[1] < intervals[index][0]: # 不需要合并区间的两种情况，1，在开头；2，在中间
            result.append(newInterval)
            result.extend(intervals[index:])
            return result
        
        begin = min(intervals[index][0], newInterval[0]) # 再合并插入的新区间
        end = intervals[index][1]
        for j in range(index, n):
            if intervals[j][1] < newInterval[1]:  # 是否需要合并区间 
                end = newInterval[1]
            elif intervals[j][0] <= newInterval[1]:  # 判断是否插入后需要再合并
                end = intervals[j][1]
            else:  # 插入之后的每一个值
                result.append([begin, end])
                begin,end = intervals[j][0], intervals[j][1]
        result.append([begin, end])
        return result
                